### PR TITLE
[#11] Add contributing guide for sgnlp

### PR DESCRIPTION
**Pull request by benedictleedm**

### CODE REVIEW

**Comment by jonheng** [comment]
The anchor links doesn't work for me

File: CONTRIBUTING.md [hunk 1]
```diff
@@ -0,0 +1,324 @@
...
+`---------------- modules (optional)`  
+`---------------- README.md`  
+`---------------- __init__.py`  
+`---------------- config.py`  
+`---------------- eval.py`  
+`---------------- modeling.py`  
+`---------------- tokenization.py (optional)`  
+`---------------- train.py`  
+`---------------- utils.py (optional)`  
+`---------------- requirements.txt (optional)`
```
Comment (jonheng) [inline CONTRIBUTING.md:62]: Project structure is a little weird. Maybe go with something like

sgnlp
|-- sgnlp
|--  |-- models

On another note, they should be forking and shouldn't need to create the entire folder structure from scratch.
So maybe can be simplified to say create a folder in sgnlp/models/<model_name> with the following files.

Overall, will prefer simplicity.

File: CONTRIBUTING.md [hunk 2]
```diff
@@ -0,0 +1,324 @@
...
+| _model_name/tokenization.py_ (optional) | Script containing the model tokenizer class which inherits from HuggingFace's [`PretrainedTokenizer`](https://huggingface.co/transformers/main_classes/tokenizer.html#transformers.PreTrainedTokenizer) class or its family of derived classes. |
+| _model_name/train.py_                   | Script containing code to train the model. It is recommended to utilize the [`Trainer`](https://huggingface.co/transformers/main_classes/trainer.html#transformers.Trainer) class from HuggingFace.                                             |
+| _model_name/README.md_                  | README markdown file containing model information such as model source, architecture, evaluation datasets and metrics, model size, and training information.                                                                                    |
+| _model_name/requirements.txt_           | Text file containing package dependencies for model serving. To manage the full list of dependencies in `sgnlp`, contributors are strongly recommended to limit their code to use the packages listed in `setup.py`.                            |
```
Comment (jonheng) [inline CONTRIBUTING.md:80]: Inconsistent in starting with sgnlp as root and describing sgnlp/sgnlp while the rest don't have the full path.
Actually with this, the above section on project structure might not be necessary. I don't think describing the root and main folder is particularly important. Can go straight to the models folder. Say it can be found in sgnlp/models.

Comment (jonheng) [reply CONTRIBUTING.md:80]: Typo in preprocess.py

Comment (jonheng) [reply CONTRIBUTING.md:80]: The requirements.txt is outdated actually. We don't want it. If there's extra dependencies involved that's out of the list indicated in setup.py, should be indicated in __init__.py

File: CONTRIBUTING.md [hunk 3]
```diff
@@ -0,0 +1,324 @@
...
+- [Submit bug fixes](#submitting-bug-fixes)
+- [Add documentation](#adding-documentation)
+
+## Adding a new model to `sgnlp`
+
+### Pre-requisites
+
+1. Fork the `sgnlp` repository.
+2. Create a Python virtual environment (verison >= 3.8)
+3. Install the packages in `requirements_dev.txt` at the root of the repository using `pip install -r requirements_dev.txt`. This will install `black`, `flake8`, and `pytest` which we use for code formatting and testing.
```
Comment (jonheng) [inline CONTRIBUTING.md:18]: Actually this is missing the local installation for the dependencies of sgnlp.
It's not this card that should do it, but the issue is how to maintain one source of truth between setup.py and requirements.txt locally.

File: CONTRIBUTING.md [hunk 4]
```diff
@@ -0,0 +1,324 @@
...
+from transformers import BertConfig
+
+class NewModelConfig(BertConfig):
+    def __init__(self, **kwargs):
+        super().__init__(**kwargs)
+```
+
+### Preprocess
+
+The `preprocess.py` script and its associated preprocessor class is an addition in `sgnlp`. When implementing various models, the team found that more complex models required more complex preprocessing. For example, some NLP models take in multiple different text inputs (ie, different utterances, multiple tweets, a single question and multiple possible answers, etc) which require different preprocessing steps. The `preprocess.py` and the preprocessor class is the team's solution to packaging all of these different steps into a single and consistent (across different models) step.
```
Comment (jonheng) [inline CONTRIBUTING.md:112]: Slight rephrasing in `...more complex models required more complex preprocessing` to `some models required more complex preprocessing`.

File: CONTRIBUTING.md [hunk 5]
```diff
@@ -0,0 +1,324 @@
...
+
+1. Check the issues list to see whether the bug has already been reported. If an issue has already been created, please comment on that issue with details on how to replicate the bug.
+2. If there is no issue relevant to the bug, please open a GitHub with
+   - Clear description of the bug
+   - Information related to your environment (ie, OS, package version, Python version, etc)
+   - Steps on how to replicate the bug (ie, a code snippet that we could run to encounter the bug)
+
+## Adding Documentation
+
+One easy way to contribute is to add or refine documentation / docstrings to the models that are currently available. `sgnlp` uses [Sphinx](https://www.sphinx-doc.org/en/master/) to generate our documentation pages. Please refer to the Sphinx documentation on the formatting expected in the docstrings. Once the docstrings have been added or edited, please submit a pull request.
```
Comment (jonheng) [inline CONTRIBUTING.md:324]: Actually the docstring format we follow is the google style python docstring. Sphinx supports this format, but I don't think Sphinx itself indicates anywhere how this format is like.

File: CONTRIBUTING.md [hunk 6]
```diff
@@ -0,0 +1,324 @@
...
+            "input_ids": torch.cat(input_ids, axis=0),
+            "attention_mask": torch.cat(attention_mask, axis=0),
+            "token_type_ids": torch.cat(token_type_ids, axis=0),
+        }
+        output = BatchEncoding(output)
+
+        return output, evidences, examples, features
+```
+
+In the RECCON Span Extraciton model, `output` is a dictionary with the token ids, attention masks and token type ids for the input utterance. `evidences`, `examples` and `features` are other features required in the RECCON model. The key idea here is to consolidate all the necessary preprocessing steps into a single method to reduce the effort needed to start using the models.
```
Comment (vincenttzc) [inline CONTRIBUTING.md:147]: Typo in Extraction

File: CONTRIBUTING.md [hunk 7]
```diff
@@ -0,0 +1,324 @@
...
+        input_ids = [torch.unsqueeze(instance[0], 0) for instance in dataset]
+        attention_mask = [torch.unsqueeze(instance[1], 0) for instance in dataset]
+        token_type_ids = [torch.unsqueeze(instance[2], 0) for instance in dataset]
+
+        output = {
+            "input_ids": torch.cat(input_ids, axis=0),
+            "attention_mask": torch.cat(attention_mask, axis=0),
+            "token_type_ids": torch.cat(token_type_ids, axis=0),
+        }
+        output = BatchEncoding(output)
```
Comment (vincenttzc) [inline CONTRIBUTING.md:142]: @raymondng76 and I used `BatchEncoding` to create the input tensors, which functions like a dictionary and a class. Should we just return the input tensors as dictionary directly instead of returning it as a BatchEncoding object? This will help to simplify things for contributors when they refer to the existing models as examples.

File: CONTRIBUTING.md [hunk 8]
```diff
@@ -0,0 +1,324 @@
...
+    logits: torch.Tensor = None
+```
+
+### train
+
+`train.py` should contain a working implementation of the model training process. A user should be able to train the model from the command line using `python -m train --train_args config/config.json`.
+
+### eval
+
+`eval.py` should contain a working implementing of the model evaluation process. This script should load the trained model (using the information in `config/config.json`) and evaluate it against the evaluation datasets. The evaluation metrics reported should correspond to that reported in the `README.md`.
```
Comment (vincenttzc) [inline CONTRIBUTING.md:288]: Think might be good to mention that there should be `train()` and `evaluate()` functions in train.py and eval.py respectively

File: CONTRIBUTING.md [hunk 9]
```diff
@@ -0,0 +1,324 @@
+# Contributing Guide
```
Comment (raymondng76) [inline CONTRIBUTING.md:1]: I think since the guide is quite long, it would be great to provide a small table of contents at the top. Something like what is described here. https://www.setcorrect.com/portfolio/work11/
